File: src/oop/shoppingList.py
from enum import Enum


class Cart:
    def __init__(self) -> None:
        self.items = dict()

    def add(self, item: str, amount: int):
        val = self.items.get(item)
        if val == None:
            self.items[item] = amount
        else:
            self.items[item] = val + amount

    def remove(self, item: str, amount: int):
        val = self.items.get(item)
        if val != None:
            if amount <= val:
                del self.items[item]
            else:
                self.items[item] = val - amount

    def remove_all(self) -> None:
        self.items = dict()


def main():
    cart: Cart = Cart()
    cart.add("Apple", 1)
    cart.add("Apple", 11)
    for k, v in cart.items.items():
        print(f"{k} : {v}")
    cart.remove("Apple", 4)
    for k, v in cart.items.items():
        print(f"{k} : {v}")
    cart.remove_all()


if __name__ == "__main__":
    main()

File: src/oop/shoppingList/shoppingList.rs
use std::collections::HashMap;

#[derive(Debug)]
struct Cart(HashMap<String, u64>); // Item name, amount
impl Cart {
    fn new() -> Self {
        Self(HashMap::<String, u64>::new())
    }
    fn add(&mut self, item: &str, amount: u64) {
        match &self.0.contains_key(&item.to_string()) {
            true => {
                *self.0.get_mut(&item.to_string()).unwrap() += amount;
            }
            false => {
                self.0.insert(item.to_string(), amount);
            }
        }
    }
    fn remove(&mut self, item: &str, amount: u64) {
        let curr_amount = *self.0.get(&item.to_string()).unwrap();
        match amount <= curr_amount {
            true => {
                *self.0.get_mut(&item.to_string()).unwrap() -= amount;
            }
            false => {
                self.0.remove(&item.to_string());
            }
        }
    }
    fn remove_all(&mut self) {
        self.0 = HashMap::<String, u64>::new();
    }
}

fn main() {
    let mut cart = Cart::new();
    cart.add("Apple", 1);
    cart.add("Apple", 10);
    cart.remove("Apple", 4);
    println!("{:?}", cart);
    cart.remove_all();
    println!("{:?}", cart);
}
